Make tieing parameters in Scatterers more transparent

barkls opened the pull request.

briandleahy commented on holopy/scattering/tests/test_composite.py:
```diff
@@ -0,0 +1,107 @@
+import unittest
```
Thanks for writing these tests. :+1:

briandleahy commented on holopy/scattering/tests/test_composite.py:
```diff
@@ -0,0 +1,107 @@
+            hp.save(tempdir + '/test.out', spheres)
+            spheres = hp.load(tempdir + '/test.out')
+        self.assertEqual(spheres.parameters.keys(), expected_keys)
+
```
I was looking through these tests to get an idea of what ```parameters``` and ```raw_parameters``` exactly returned, and I noticed that ```raw_parameters``` isn't tested anywhere. Can we add a unit test for ```raw_parameters```?

Also, if ```raw_parameters``` is not public, it might make sense to make that a little clearer (e.g. rename it ```_raw_parameters``` if outside class don't need to use it).

briandleahy commented on holopy/scattering/scatterer/composite.py:
```diff
@@ -74,45 +85,71 @@ def get_component_list(self):
+        for tied_name, raw_names in self.ties.items():
+            for raw_name in raw_names:
+                if raw_name not in parameters.keys():
+                    msg = 'Tied parameter {} not present in raw \
```
When does this (and the other) error message get raised in practice? It seems that they are raised if something is not created correctly, i.e. only if there is a bug in the code here. If that's the case I think it's great to have them, but for readability you might want to make them ```assert```s rather than ```ValueError```s. If they are possible to raise by the user, then it would be nice to have some tests (at least so other developers can see what you are checking for and what fail cases are).

briandleahy commented on holopy/scattering/scatterer/composite.py:
```diff
@@ -74,45 +85,71 @@ def get_component_list(self):
-        for i, scatterer in enumerate(self.scatterers):
-            for key, par in scatterer.parameters.items():
-                for index, par_check in enumerate(pars + [None]):
+    def update_ties(self):
```
As a general comment / question:

This seems like some pretty powerful stuff here. Is there a reason it's limited to ```Scatterers``` and not to the more general ```Scatterer``` class? For instance, I could imagine wanting to fit an ellipsoid with the two minor axes equal.

If this is not a very easy fix then let's not do it for this PR and do it for a later update when desired.

barkls replied:
It's not a trivially easy change, but it wouldn't be too hard to do. I just can't think of an obvious use case. In your example, you should just use the `spheroid` class instead of `ellipsoid`.

briandleahy replied:
Let's just leave it as is then for now.

briandleahy commented on holopy/scattering/scatterer/composite.py:
```diff
@@ -74,45 +85,71 @@ def get_component_list(self):
+                    msg = 'Tied parameters {} and {} are not equal.'.format(
+                            parameters[raw_name], tied_val)
+                    raise ValueError(msg)
+                del parameters[raw_name]
```
OK. I'm a little confused what this function is doing. It seems that you are 
1. Copying ```raw_parameters```, which contains duplicates.
2. Removing a parameter one time when it appears in ```self.ties```.
I'm confused as to why this works, as it seems to me that this assumes that tied parameters only appear in ```raw_parameters``` two times at most (e.g. I think this will fail if there are 3 spheres all tied to share the same index). If that's the case, why doesn't something like this work?
```python
@property
def parameters(self):
    ties_reversed = {raw_name: tied_name for tied_name, raw_name in self.ties.items()}
    parameters = {
        name: parameter for name, parameter in self.raw_parameters.items()
        if name not in ties_reversed else ties_reversed[name]: parameter}
    return parameters
```

Or am I missing something?

barkls commented:
I added more tests to make sure I have 100% coverage of the new code I wrote. 

I refactored a few things to remove some recursive calls and get rid of the `del` statement.

I also broke out a new method `_check_ties` since `parameters` was doing multiple different things.

Would appreciate another look from @briandleahy when you get a chance!

briandleahy commented:
Hey @barkls thanks for fixing these! The added tests and documentation look great!

I'm still a little confused about some of the actual code though. Can we speak tomorrow morning, say 9am-ish, about it?

barkls commented:
Yes

On Mon, Aug 19, 2019, 17:01 Brian Leahy <notifications@github.com> wrote:

> Hey @barkls <https://github.com/barkls> thanks for fixing these! The
> added tests and documentation look great!
>
> I'm still a little confused about some of the actual code though. Can we
> speak tomorrow morning, say 9am-ish, about it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/manoharan-lab/holopy/pull/290?email_source=notifications&email_token=AETUFJZLYZVHMRNESNY4VGTQFMCYXA5CNFSM4IMLQSG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4UJAAQ#issuecomment-522752002>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AETUFJ65OW6PDKVLCOIJ6L3QFMCYXANCNFSM4IMLQSGQ>
> .
>

barkls commented:
I didn't check it compared to any other HoloPy operations, but I did
compare against an alternative way that turned out to be much slower so I
left as is.

I'm ready for merge.

On Tue, 20 Aug 2019 at 15:38, Brian Leahy <notifications@github.com> wrote:

> *@briandleahy* approved this pull request.
>
> Looks good to me. I'm ready to merge.
>
> Did you profile these and see if the from_parameters method was slow? If
> it's not slow, let's merge it. If it is slow, let's talk about it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/manoharan-lab/holopy/pull/290?email_source=notifications&email_token=AETUFJ5DXEKLRY7B4L25TIDQFRB27A5CNFSM4IMLQSG2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCELJ2Q#pullrequestreview-277394666>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AETUFJ2LSJEPOIQWYJOSPITQFRB27ANCNFSM4IMLQSGQ>
> .
>
